### PR TITLE
Update UpdatePolicyConnectionConfiguration to be 1:1:1 Part-3

### PR DESCRIPTION
**Pull request by ZhenjaHorbach**
### Details

Update UpdatePolicyConnectionConfiguration to be 1:1:1 - Part 3

### Fixed Issues

$ https://github.com/Expensify/App/issues/47521
PROPOSAL: NA

### Tests

1. Open App
2. Open Account settings and then Workspaces
3. Choose any Workspace that has Quickbooks Online connection
4. Open Accounting 
5. Open Import in Quickbooks Online
6. Choose classes 
7. **Verify that Import can be turned on and off and the request is sent correctly without errors**
8. Press go back
9. Choose Customers/projects
10. **Verify that Import can be turned on and off and the request is sent correctly without errors**
11. Press go back
12. Choose Locations
13. **Verify that Import can be turned on and off and the request is sent correctly without errors**
14. Press go back
15. Choose Taxes
16. **Verify that Import can be turned on and off and the request is sent correctly without errors**
17. Return to Accounting screen
18. Open Export in Quickbooks Online
19. Open Export company card expenses as and choose Vendor bill in Export as
20. Open Default vendor
21. **Verify that vendor can be selected and the request is sent correctly without errors**

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/bb300b9e-eed3-4096-8073-b141abb03bdb

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/3d357d5d-27f8-404f-8c76-ac4ce75e1ff3

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/c1ca7784-f9bc-4f7f-849b-b17d34dab72d

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/bc08c290-6a7b-4dbe-80d3-e7af176c3a1a

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/2e003b59-4ffa-4b8c-9b79-8e7905bed330

https://github.com/user-attachments/assets/2d7b8fc7-8dec-4dd3-bea6-9fb7f2a88362

https://github.com/user-attachments/assets/d41a1513-75f0-4a45-bf61-38122d1c70d5

https://github.com/user-attachments/assets/94f93764-91d0-4214-8048-2c54efa23021

https://github.com/user-attachments/assets/3b26f81a-a33a-41dd-933e-b3ddfeda37aa

https://github.com/user-attachments/assets/f3c9e8bd-df92-4dfa-9f74-8574b6e601f0

https://github.com/user-attachments/assets/634dad95-02bc-4c13-9259-cc95aa986ac2

https://github.com/user-attachments/assets/52ff09b7-d250-4434-a113-757e1b5f1793

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/914ec0cb-cda4-42c0-a645-3b9d786c9492

</details>

### CODE REVIEW

**Review by ishpaul777 — APPROVED** [review]
LGTM!

**Review by dangrous — CHANGES_REQUESTED** [review]
This looks good! Love the simplification of it. Do you want to go ahead and update the other ones to that format too? EnableNewCategories, AutoCreateVendor, ReimbursableExpensesAccount?

Also @aldo-expensify was AutoCreateVendor the one we needed to update into the new format with two params? I forget.

**Comment by aldo-expensify** [comment]
> Also @aldo-expensify was AutoCreateVendor the one we needed to update into the new format with two params? I forget.

Yep, `UpdateQuickbooksOnlineAutoCreateVendor` needs the following params:

`autoCreateVendor`: boolean
`nonReimbursableBillDefaultVendor`: string

I think it should work like what we have here:

https://github.com/Expensify/App/blob/1bb113191ddf05810e795e04f7651ddc89a58d5a/src/pages/workspace/accounting/qbo/export/QuickbooksCompanyCardExpenseAccountPage.tsx#L73-L87

**Comment by dangrous** [comment]
oh yeah so let's update that ^^ call to `updateManyPolicyConnectionConfigs` as well!

Sorry to throw extra work on you @ZhenjaHorbach

**Comment by dangrous** [comment]
Oops sorry I phrased that poorly. I meant - can we change the previous call to `updateManyPolicyConnectionConfigs` to call instead `UpdateQuickbooksOnlineAutoCreateVendor` with both parameters (`autoCreateVendor` and `nonReimbursableBillDefaultVendor`). It will accept it on the backend.

And @aldo-expensify I believe you were saying that we needed to update [where we call autoCreateVendor with only true/false](https://github.com/Expensify/App/blob/f04c8883f3859f930c3c226dca9b5c9cecb1a72d/src/pages/workspace/accounting/qbo/advanced/QuickbooksAdvancedPage.tsx#L103) to also send a `nonReimbursableBillDefaultVendor`? That it was an existing bug? If so (or if not) can you clarify how we should go about that one?

**Comment by dangrous** [comment]
oh so @ZhenjaHorbach I think you basically did the right thing, but instead of UpdateMany... we're using AutoCreateVendor but with two parameters.

**Comment by dangrous** [comment]
oh got it yeah we have it as separate params - so instead of `configUpdate` do individually `autoCreateVendor` and `nonReimbursableBillDefaultVendor`

**Comment by aldo-expensify** [comment]
> And @aldo-expensify I believe you were saying that we needed to update [where we call autoCreateVendor with only true/false](https://github.com/Expensify/App/blob/f04c8883f3859f930c3c226dca9b5c9cecb1a72d/src/pages/workspace/accounting/qbo/advanced/QuickbooksAdvancedPage.tsx#L103) to also send a nonReimbursableBillDefaultVendor? That it was an existing bug? If so (or if not) can you clarify how we should go about that one?

Yes, I think you have to do it similar to the existing case where we update `nonReimbursableBillDefaultVendor` ([here](https://github.com/Expensify/App/blob/1bb113191ddf05810e795e04f7651ddc89a58d5a/src/pages/workspace/accounting/qbo/export/QuickbooksCompanyCardExpenseAccountPage.tsx#L73-L87)):

If you are **enabling** `autoCreateVendor`, you should change `nonReimbursableBillDefaultVendor` to the first element found in `policy.connections.quickbooksOnline.data.vendors[0].id`, something like this:

```
policy?.connections?.quickbooksOnline?.data?.vendors?.[0]?.id ?? CONST.INTEGRATION_ENTITY_MAP_TYPES.NONE 
```

if you are **disabling** `autoCreateVendor`, you should change `nonReimbursableBillDefaultVendor` to `CONST.INTEGRATION_ENTITY_MAP_TYPES.NONE`

**Review by ishpaul777 — APPROVED** [review]
LGTM!

**Comment by aldo-expensify** [comment]
you have some "React Compiler" error:

```

Error: Some files could be compiled with react-compiler before successfully, but now they can not be compiled. Check https://github.com/Expensify/App/blob/main/contributingGuides/REACT_COMPILER.md documentation to see how you can fix this.
```

There are some instructions on how to run it locally and find the cause of the problem: https://github.com/Expensify/App/blob/main/contributingGuides/REACT_COMPILER.md

**Comment by aldo-expensify** [comment]
> I thought we are ignoring this [#47628 (comment)](https://github.com/Expensify/App/pull/47628#issuecomment-2317198112) but yeah if we can fix that would be great

ahh I wasn't aware 👍

File: src/libs/actions/connections/QuickbooksOnline.ts [hunk 1]
```diff
@@ -21,7 +22,11 @@ function getQuickbooksOnlineSetupLink(policyID: string) {
     return commandURL + new URLSearchParams(params).toString();
 }
 
-function updateQuickbooksOnlineEnableNewCategories(policyID: string, settingValue: boolean) {
+function updateManyQuickbooksOnyxData<TConfigUpdate extends Partial<Connections['quickbooksOnline']['config']>>(
```
Comment (dangrous) [inline src/libs/actions/connections/QuickbooksOnline.ts:25]:
```suggestion
function buildOnyxDataForMultipleQuickbooksConfigurations<TConfigUpdate extends Partial<Connections['quickbooksOnline']['config']>>(
```

NAB! This is also not a good name but is a little clearer?

**Review by dangrous — DISMISSED** [review]
this looks good! One potential rename but otherwise works for me. @aldo-expensify do you want to give this one a review as well?

**Comment by aldo-expensify** [comment]
> @aldo-expensify do you want to give this one a review as well?

Sure, I'll review asap

File: src/libs/actions/connections/QuickbooksOnline.ts [hunk 2]
```diff
@@ -21,7 +22,11 @@ function getQuickbooksOnlineSetupLink(policyID: string) {
     return commandURL + new URLSearchParams(params).toString();
 }
 
-function updateQuickbooksOnlineEnableNewCategories(policyID: string, settingValue: boolean) {
+function buildOnyxDataForMultipleQuickbooksConfigurations<TConfigUpdate extends Partial<Connections['quickbooksOnline']['config']>>(
```
Comment (aldo-expensify) [inline src/libs/actions/connections/QuickbooksOnline.ts:25]: Nice, I see you were able to avoid a lot of code repetition by creating this helper method

**Review by aldo-expensify — DISMISSED** [review]
Looks great, thanks!

**Review by dangrous — DISMISSED** [review]
LGTM! @aldo-expensify do you know what we can do about ignoring that compiler check? Just merge anyway and call it out in a comment?

**Comment by aldo-expensify** [comment]
> Just merge anyway and call it out in a comment?

Yes, that is my understanding. I think an "Emergency" label will be added and we have to remove it and comment why.

**Review by dangrous — APPROVED** [review]

**Review by aldo-expensify — APPROVED** [review]

**Comment by dangrous** [comment]
Removing label - react-compiler check is known and is not a blocker - see https://github.com/Expensify/App/pull/47628#issuecomment-2326722367

File: src/libs/actions/connections/QuickbooksOnline.ts [hunk 3]
```diff
@@ -127,12 +123,12 @@ function updateQuickbooksOnlineAutoCreateVendor(policyID: string, settingValue:
                 connections: {
                     [CONST.POLICY.CONNECTIONS.NAME.QBO]: {
                         config: {
-                            [CONST.QUICK_BOOKS_CONFIG.AUTO_CREATE_VENDOR]: settingValue ?? null,
+                            [settingName]: settingValue ?? null,
```
[the remark this answers is not in the document]
Comment (dangrous) [reply src/libs/actions/connections/QuickbooksOnline.ts:126]: @aldo-expensify you're working on QBO errors right? I'd defer to you for this one. My naive opinion is show the error text and revert to the previous value (so, both?).

[the remark this answers is not in the document]
Comment (aldo-expensify) [reply src/libs/actions/connections/QuickbooksOnline.ts:126]: > @aldo-expensify you're working on QBO errors right?

I'm not, but this does looks like an oversight. `buildOnyxDataForMultipleQuickbooksConfigurations` is reverting to the original value, why wouldn't `buildOnyxDataForQuickbooksConfiguration` do the same.

> My naive opinion is show the error text and revert to the previous value (so, both?).

Agree

[the remark this answers is not in the document]
Comment (war-in) [reply src/libs/actions/connections/QuickbooksOnline.ts:126]: Hi 👋 @zfurtak is working on the QBO offline and errors refactor [here](https://github.com/Expensify/App/pull/46511) so let's wait for her opinion 🙏

[the remark this answers is not in the document]
Comment (zfurtak) [reply src/libs/actions/connections/QuickbooksOnline.ts:126]: Hi guys 😊 The new strategy for errors that we were introducing is that after an error appears the value is reverting to the previous one.
If you want to, take a look into [my PR](https://github.com/Expensify/App/pull/46511/files#diff-7b811795926d7001296df148e05cadd46ac9cf78bdb6418d2787d1172b8e237b) where I adjusted already changed functions to this pattern 😉

[the remark this answers is not in the document]
Comment (dangrous) [reply src/libs/actions/connections/QuickbooksOnline.ts:126]: oh okay great! @ZhenjaHorbach can you take a look at that and update your update PR accordingly? Thank you!

[the remark this answers is not in the document]
Comment (zfurtak) [reply src/libs/actions/connections/QuickbooksOnline.ts:126]: Hello 👋 FYI my PR was merged 😃

**Pull request merged**